C++/Navio/Common/gpio.cpp

```diff
@@ -137,37 +132,48 @@ void Pin::toggle()
     write(!read());
 }
 
+uint32_t Pin::getPeriBase() const
+{
+    int rpi_version = getRaspberryPiVersion();
+
+    switch (rpi_version) {
+      case 1:
+        return BCM2708_PERI_BASE;
+      case 2:
+        return BCM2709_PERI_BASE;
+      case 3:
+        return BCM2835_PERI_BASE;
+      default:
+        fprintf(stderr, "Unexpected version. Setting up configurations as on RPI2/3\n");
+        return BCM2709_PERI_BASE;
+}
+}
+
+
+
 int Pin::getRaspberryPiVersion() const
 {
-    char buffer[MAX_SIZE_LINE];
-    const char* hardware_description_entry = "Hardware";
-    const char* v1 = "BCM2708";
-    const char* v2 = "BCM2709";
-    const char* v3 = "BCM2835";
-    char* flag;
-    FILE* fd;
-
-    fd = fopen("/proc/cpuinfo", "r");
-
-    while (fgets(buffer, MAX_SIZE_LINE, fd) != NULL) {
-        flag = strstr(buffer, hardware_description_entry);
-
-        if (flag != NULL) {
-            if (strstr(buffer, v3) != NULL) {
-                fclose(fd);
-                return 3;
-            } else if (strstr(buffer, v2) != NULL) {
-                fclose(fd);
-                return 2;
-            } else if (strstr(buffer, v1) != NULL) {
-                fclose(fd);
-                return 1;
-            }
+    int rpi_version;
+
+    size_t i = 0;
+    std::string buffer;
+    std::ifstream f(SYSFS_RPI_MODEL_FILE_PATH);
+
+    if (f.is_open()){
+      while (std::getline(f, buffer)) {
+        while (!isdigit(buffer[i])) { //first digit is a number of RPI version (from 1 to 4)
+          i++; //skip
```

Comment:
This is bad. There can be _no_ digits in the common case. we do not know this. Adding a potentially infinite loop is a bad decision.

Reply:
Can I not use `break` doing it like this:
`while (i < buffer.length()) {
          if (isdigit(buffer[i])) {
            char version_char = buffer[i];
            rpi_version = atoi(&version_char);

            return rpi_version;
          }
          i++;
        }`

Reply:
So, it looks for loop, isn't it?

Reply:
For sure, and without `break`